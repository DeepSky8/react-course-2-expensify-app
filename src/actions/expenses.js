import { db } from '../firebase/firebase';
import { get, push, ref, remove, update } from 'firebase/database';

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
})

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            amount = 0,
            createdAt = 0,
            description = '',
            note = ''
        } = expenseData
        const expense = {
            amount,
            createdAt,
            description,
            note
        }

        return push(ref(db, 'expenses'), expense)
            .then((ref) => {
                dispatch(addExpense({
                    id: ref.key,
                    ...expense
                }))
            })
            .catch((e) => { })
    }
}

export const removeExpense = ({ id } = {}) => (
    {
        type: 'REMOVE_EXPENSE',
        id
    }
);



export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch) => {
        return remove(ref(db, `expenses/${id}`))
            .then(() => {
                dispatch(removeExpense({ id }))
            })
            .catch((e) => {
                console.log('Error: ', e)
            })
    }
}

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch) => {
        return get(ref(db, 'expenses'))
            .then((dataSnapshot) => {
                const expenses = [];
                dataSnapshot.forEach((childSnapShot) => {
                    expenses.push({
                        id: childSnapShot.key,
                        ...childSnapShot.val()
                    })
                })
                dispatch(setExpenses(expenses))
            }).catch((error) => {
                console.log('Error: ', error)
            });
    }
};



