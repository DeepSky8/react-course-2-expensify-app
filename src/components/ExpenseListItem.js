import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
    <div>
        <nav>
            <Link to={`edit/${id}`}>
                <h3>{description}</h3>
            </Link>
        </nav>

        <p>Amount: {amount}</p>
        <p>Created: {createdAt}</p>
    </div>
)


export default ExpenseListItem;