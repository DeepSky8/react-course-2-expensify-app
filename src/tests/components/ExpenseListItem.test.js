import React from 'react';
import ReactShallowRenderer from 'react-test-renderer/shallow';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';

test('should render ExpenseListItem with an expense', () => {
    const renderer = new ReactShallowRenderer();
    renderer.render(<ExpenseListItem {...expenses[0]} />)
    expect(renderer.getRenderOutput()).toMatchSnapshot();
})