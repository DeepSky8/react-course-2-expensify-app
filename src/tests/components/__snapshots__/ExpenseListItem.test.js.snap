// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with an expense 1`] = `
<div>
  <nav>
    <Link
      to="edit/1"
    >
      <h3>
        Gum
      </h3>
    </Link>
  </nav>
  <p>
    Amount: 
    $1.95
  </p>
  <p>
    Created: 
    December 31st 1969
  </p>
</div>
`;
